src/providers/activeLearning/objectDetection.ts

```diff
@@ -0,0 +1,195 @@
+import axios from "axios";
+import * as tf from "@tensorflow/tfjs";
+import { ElectronProxyHandler } from "./electronProxyHandler";
+import { LocalFileSystemProxy, ILocalFileSystemProxyOptions } from "../../providers/storage/localFileSystemProxy";
+
+// tslint:disable-next-line:interface-name
+export interface DetectedObject {
+    bbox: [number, number, number, number];  // [x, y, width, height]
+    class: string;
+    score: number;
+}
+
+export class ObjectDetection {
+    private modelLoaded: boolean = false;
+
+    get loaded(): boolean {
+        return this.modelLoaded;
+    }
+
+    private model: tf.GraphModel;
+    private jsonClasses: JSON;
+
+    /**
+     * Dispose the tensors allocated by the model. You should call this when you
+     * are done with the model.
+     */
+    public dispose() {
+        if (this.model) {
+            this.model.dispose();
+        }
+    }
+
+    public async load(modelFolderPath: string) {
+        try {
+            if (modelFolderPath.toLowerCase().startsWith("http://") ||
+                modelFolderPath.toLowerCase().startsWith("https://")) {
+                this.model = await tf.loadGraphModel(modelFolderPath + "/model.json");
+
+                const response = await axios.get(modelFolderPath + "/classes.json");
+                this.jsonClasses = JSON.parse(JSON.stringify(response.data));
+            } else {
+                if (modelFolderPath.toLowerCase().startsWith("file://")) {
+                    modelFolderPath = modelFolderPath.substring(7);
+                }
+                const handler = new ElectronProxyHandler(modelFolderPath);
+                this.model = await tf.loadGraphModel(handler);
+
+                const provider = new LocalFileSystemProxy();
+                this.jsonClasses = JSON.parse(await provider.readText(modelFolderPath + "/classes.json"));
+            }
+
+            // Warmup the model.
+            const result = await this.model.executeAsync(tf.zeros([1, 300, 300, 3])) as
+                tf.Tensor[];
+            result.forEach(async (t) => await t.data());
+            result.forEach(async (t) => t.dispose());
+            this.modelLoaded = true;
+        } catch (_) {
+            this.modelLoaded = false;
+        }
+    }
+
+    /**
+     * Detect objects for an image returning a list of bounding boxes with
+     * assocated class and score.
+     *
+     * @param img The image to detect objects from. Can be a tensor or a DOM
+     *     element image, video, or canvas.
+     * @param maxNumBoxes The maximum number of bounding boxes of detected
+     * objects. There can be multiple objects of the same class, but at different
+     * locations. Defaults to 20.
+     *
+     */
+    public async detect(img: tf.Tensor3D|ImageData|HTMLImageElement|HTMLCanvasElement|HTMLVideoElement,
```

Comment:
Consider declaring a typescript `type` that includes all these other type options.